staging/src/k8s.io/client-go/tools/watch/retrywatcher.go

```diff
@@ -0,0 +1,232 @@
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog"
+)
+
+// WatcherFunc is a function that is responsible for creating a watcher starting at sinceResourceVersion.
+type WatcherFunc func(sinceResourceVersion string) (watch.Interface, error)
+
+// resourceVersionGetter is an interface used to get resource version from events.
+// We can't reuse an interface from meta otherwise it would be a cyclic dependency and we need just this one method
+type resourceVersionGetter interface {
+	GetResourceVersion() string
+}
+
+// RetryWatcher will make sure that in case the underlying watcher is closed (e.g. due to API timeout or etcd timeout)
+// it will get restarted from the last point without the consumer even knowing about it.
+// RetryWatcher does that by inspecting events and keeping track of resourceVersion.
+// Especially useful when using watch.Until where premature termination is causing issues and flakes.
+// Please note that this is not resilient to etcd cache not having the resource version anymore - you would need to
+// use Informers for that.
+type RetryWatcher struct {
+	lastResourceVersion string
+	watcherFunc         WatcherFunc
+	resultChan          chan watch.Event
+	stopChan            chan struct{}
+	doneChan            chan struct{}
+}
+
+// NewRetryWatcher creates a new RetryWatcher.
+// It will make sure that watcher gets restarted in case of recoverable errors.
+// The initialResourceVersion will be given to watchFunc when first called.
+func NewRetryWatcher(initialResourceVersion string, watcherFunc WatcherFunc) *RetryWatcher {
+	rw := &RetryWatcher{
+		lastResourceVersion: initialResourceVersion,
+		watcherFunc:         watcherFunc,
+		stopChan:            make(chan struct{}),
+		doneChan:            make(chan struct{}),
+		resultChan:          make(chan watch.Event, 0),
+	}
+
+	go rw.receive()
+	return rw
+}
+
+func (rw *RetryWatcher) send(event watch.Event) bool {
+	// Writing to an unbuffered channel is blocking operation
+	// and we need to check if stop wasn't requested while doing so.
+	select {
+	case rw.resultChan <- event:
+		return true
+	case <-rw.stopChan:
+		return false
+	}
+}
+
+// doReceive returns true when it is done, false otherwise.
+// If it is not done the second return value holds the time to wait before calling it again.
+func (rw *RetryWatcher) doReceive() (bool, time.Duration) {
+	watcher, err := rw.watcherFunc(rw.lastResourceVersion)
+	if err != nil {
+		klog.Errorf("retryWatcher: watcherFunc failed: %v", err)
+		// Retry
+		return false, 0
+	}
+
+	if watcher == nil {
+		klog.Error("retryWatcher: watcherFunc returned nil watcher")
+		// Retry
+		return false, 0
+	}
+
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+
+	for {
+		select {
+		case <-rw.stopChan:
+			klog.V(4).Info("Stopping RetryWatcher.")
+			return true, 0
+		case event, ok := <-ch:
+			if !ok {
+				klog.Warningf("RetryWatcher - getting event failed! Re-creating the watcher. Last RV: %s", rw.lastResourceVersion)
```

Comment:
isn't the channel closing a normal occurrence? warning level logging seems way too high... I'd have thought v4 or so

Reply:
you are right